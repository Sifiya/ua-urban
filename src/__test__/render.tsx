import React, { ReactNode } from 'react';
import { render as libRender } from '@testing-library/react';
import { Providers } from '@/components/Providers';

const Wrapper = ({ children }: { children: ReactNode }) => {
  return <Providers isTest>{children}</Providers>;
};

export const render = (ui: ReactNode) => {
  return libRender(ui, {
    wrapper: Wrapper,
  });
};
