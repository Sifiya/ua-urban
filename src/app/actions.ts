'use server';
import type { Database } from '@/types/supabase';
import type { Word, Definition } from '@/types/types';
import { createClient } from '@supabase/supabase-js';

const supabase = createClient<Database>(
  process.env.SUPABASE_API_URL || '',
  process.env.SUPABASE_API_KEY || ''
);

export const addWordWithDefinition = async (word: string, definition: string) => {
  try {
    const createdWord: Word = await addWord(word);
    await addDefinition(definition, createdWord.id);
  } catch (error) {
    return { error, success: false };
  }
  return { error: null, success: true };
};

const addWord = async (word: string): Promise<Word> => {
  const { data: createdWord, error } = await supabase.from('words').insert([
    { word },
  ]).select();

  if (error || !createdWord) {
    throw error || new Error('Не вдалося створити слово');
  }

  return createdWord[0];
};

const addDefinition = async (definition: string, wordId: string): Promise<Definition> => {
  const { data: createdDefinition, error } = await supabase.from('definitions').insert([
    { text: definition, word_id: wordId },
  ]).select();

  if (error || !createdDefinition) {
    throw error || new Error('Не вдалося створити визначення');
  }

  return createdDefinition[0];
};

export const getAllWords = async (): Promise<Word[]> => {
  const { data: words, error } = await supabase.from('words').select('*');

  if (error || !words) {
    throw error || new Error('Не вдалося отримати слова');
  }

  return words;
};
