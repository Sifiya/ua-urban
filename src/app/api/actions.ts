'use server';
import type { Word, MinimalWord, Definition, MinimalDefinition } from '@/types/types';
import { supabase } from './supabase.init';

export const addWordWithDefinition = async (word: string, definition: string) => {
  let createdWord: Word;
  try {
    createdWord = await addWord(word);
    await addDefinition(definition, createdWord.id);
  } catch (error) {
    return { error, success: false };
  }
  return { error: null, success: true, word_id: createdWord.id };
};

const addWord = async (word: string): Promise<Word> => {
  const { data: createdWord, error } = await supabase.from('words').insert([
    { word },
  ]).select();

  if (error || !createdWord) {
    throw error || new Error('Не вдалося створити слово');
  }

  return createdWord[0];
};

export const addDefinition = async (definition: string, wordId: string): Promise<Definition> => {
  const { data: createdDefinition, error } = await supabase.from('definitions').insert([
    { text: definition, word_id: wordId },
  ]).select();

  if (error || !createdDefinition) {
    throw error || new Error('Не вдалося створити визначення');
  }

  return createdDefinition[0];
};

export const getAllWords = async (): Promise<Word[]> => {
  const { data: words, error } = await supabase.from('words').select('*');

  if (error || !words) {
    throw error || new Error('Не вдалося отримати слова');
  }

  return words;
};

export const getOneWord = async (id: string): Promise<MinimalWord> => {
  const { data: words, error } = await supabase.from('words').select('id,word').eq('id', id);

  if (error || !words || words.length === 0) {
    throw error || new Error('Не вдалося отримати слова');
  }

  return words[0];
};

export const getWordDefinitions = async (wordId: string): Promise<MinimalDefinition[]> => {
  const { data: definitions, error } = await supabase.from('definitions').select('id,word_id,text').eq('word_id', wordId);

  if (error || !definitions) {
    throw error || new Error('Не вдалося отримати визначення');
  }

  return definitions;
};
