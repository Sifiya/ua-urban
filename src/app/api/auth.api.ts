'use server';
import { createClient } from '@/utils/supabase/server';

const supabase = createClient();

const withErrorHandling = <T, K>(data?: K, error?: T): {
  error: T | null,
  success: boolean,
  data?: K,
} => {
  if (error || !data) {
    return { error: error || null, success: false };
  }

  return { error: null, success: true, data};
};

export const signUp = async (email: string, password: string) => {
  const { data: user, error } = await supabase.auth.signUp({
    email,
    password,
  });

  return withErrorHandling(user, error?.message);
};

export const signInWithEmail = async (email: string, password: string) => {
  const { data, error } = await supabase.auth.signInWithPassword({
    email,
    password,
  });

  return withErrorHandling(data.session?.access_token, error?.message);
};

export async function signOut() {
  const { error } = await supabase.auth.signOut()
  return withErrorHandling(true, error);
}
