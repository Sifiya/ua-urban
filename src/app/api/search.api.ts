'use server';
import { createClient } from '@/utils/supabase/server';

export const searchWord = async (searchString: string, noLimit: boolean = false) => {
  const supabase = createClient();
  let query = supabase
    .from('words')
    .select('id, word')
    .ilike('word', `%${searchString}%`);

  if (!noLimit) {
    query = query.limit(10);
  }
    
  query = query.order('word', { ascending: true });

  const { data, error } = await query;
  if (error) {
    throw error;
  }
  return data;
};
