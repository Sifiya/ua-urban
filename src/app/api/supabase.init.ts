import { createClient } from '@supabase/supabase-js';
import type { Database } from '@/types/supabase';

export const supabase = createClient<Database>(
  process.env.SUPABASE_API_URL || '',
  process.env.SUPABASE_API_KEY || ''
);
