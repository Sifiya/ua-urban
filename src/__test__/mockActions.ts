import * as actions from '@/app/api/actions';
import * as authApi from '@/app/api/auth.api';

jest.mock('@/app/api/actions', () => ({
  addWordWithDefinition: jest.fn(),
  getAllWords: jest.fn(),
  addDefinition: jest.fn(),
}));

export const mockAddWordWithDefinition = actions.addWordWithDefinition as jest.Mock;
export const mockGetAllWords = actions.getAllWords as jest.Mock;
export const mockAddDefinition = actions.addDefinition as jest.Mock;

jest.mock('@/app/api/auth.api', () => ({
  getUser: jest.fn(),
}));

export const mockGetUser = authApi.getUser as jest.Mock;
