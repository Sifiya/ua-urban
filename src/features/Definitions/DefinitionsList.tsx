'use client';

import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { getWordDefinitions } from '@/app/api/actions';
import { useProfile } from '@/hooks/useProfile';
import { Card } from '@/components/ui/card';
import GridLoader from 'react-spinners/GridLoader';
import { VoteBlock } from './VoteBlock';

interface DefinitionsListProps {
  wordId: string;
}

export const DefinitionsList = ({ wordId }: DefinitionsListProps) => {
  const { isAuthenticated } = useProfile()
  const { data: definitions = [], isLoading } = useQuery({
    queryKey: ['word', wordId, 'definitions'],
    queryFn: () => getWordDefinitions(wordId),
  });

  if (isLoading) {
    return (
      <div className="w-full flex justify-center p-10">
        <GridLoader size={15} color="#435f69" />
      </div>
    );
  }

  return definitions.map(({ id, text }) => (
    <Card key={id} className="py-3 px-5">
      <p>
        {text}
      </p>
      {isAuthenticated && (<VoteBlock definitionId={id} wordId={wordId} />)}
    </Card>
  ));
};
