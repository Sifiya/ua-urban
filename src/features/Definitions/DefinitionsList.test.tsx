import React from 'react';
import { screen } from '@testing-library/react';
import { render } from '@/__test__/render';
import * as actions from '@/app/api/actions';
import { DefinitionsList } from './DefinitionsList';

jest.mock('./VoteBlock', () => ({
  VoteBlock: () => <div>VoteBlock</div>,
}));
jest.mock('@/app/api/actions', () => ({
  getWordDefinitions: jest.fn(),
}));
const getWordDefinitionsMock = actions.getWordDefinitions as jest.Mock;

describe('DefinitionsList', () => {
  afterEach(() => {
    getWordDefinitionsMock.mockReset();
  });

  test('should render definitions list', async () => {
    const definitions = [
      { id: '1', text: 'Definition 1', upvotes_count: 2, downvotes_count: 1 },
      { id: '2', text: 'Definition 2', upvotes_count: 3, downvotes_count: 0 },
    ];
    getWordDefinitionsMock.mockResolvedValue(definitions);

    const { container } = render(<DefinitionsList wordId="1" />);

    expect(await screen.findByText('Definition 1')).toBeInTheDocument();
    expect(screen.getByText('Definition 2')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
