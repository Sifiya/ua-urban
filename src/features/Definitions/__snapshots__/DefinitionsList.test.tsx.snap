// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefinitionsList should render definitions list 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm py-3 px-5"
  >
    <div
      class="p-6 pt-0"
    >
      <p
        class="leading-7 [&:not(:first-child)]:mt-6 undefined"
      >
        Definition 2
      </p>
      <div>
        VoteBlock
      </div>
    </div>
  </div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm py-3 px-5"
  >
    <div
      class="p-6 pt-0"
    >
      <p
        class="leading-7 [&:not(:first-child)]:mt-6 undefined"
      >
        Definition 1
      </p>
      <div>
        VoteBlock
      </div>
    </div>
  </div>
</div>
`;
